protocol/msgpack/base_iterator.go

```diff
@@ -61,15 +66,40 @@ func (it *baseIterator) err() error       { return it.decodeErr }
 func (it *baseIterator) setErr(err error) { it.decodeErr = err }
 
 func (it *baseIterator) decodePolicy() policy.Policy {
-	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(policyType)
+	numActualFields := it.decodeNumObjectFields()
+	policyType := it.decodeObjectType()
+
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
```

Comment:
Seems like an odd function - takes in numActualFields and gets numActualFields returned. Do they differ in any way? Why not have this function return you the delta and you use the delta below on line 85 and 97